sdk/core/Azure.Core/tests/OperationInternalTests.cs

```diff
@@ -0,0 +1,532 @@
+                    _ => OperationState<int>.Success(new MockResponse(200), 50)
+                }
+            };
+            var operationInternal = testOperation.MockOperationInternal;
+
+            operationInternal.OnWait = delay => passedDelays.Add(delay);
+
+            if (useDefaultPollingInterval)
+            {
+                operationInternal.DefaultPollingInterval = originalDelay;
+                await operationInternal.WaitForCompletionAsync(CancellationToken.None);
+            }
+            else
+            {
+                await operationInternal.WaitForCompletionAsync(originalDelay, CancellationToken.None);
+            }
+
+            // Algorithm must choose the longest delay between the two.
+            Assert.AreEqual(2, passedDelays.Count);
+            Assert.AreEqual(originalDelay, passedDelays[0]);
+            Assert.AreEqual(longDelay, passedDelays[1]);
+        }
+
+        [Test]
+        [TestCase(true, "retry-after-ms")]
+        [TestCase(true, "x-ms-retry-after-ms")]
+        [TestCase(false, "retry-after-ms")]
+        [TestCase(false, "x-ms-retry-after-ms")]
+        public async Task WaitForCompletionUsesRetryAfterMsHeader(bool useDefaultPollingInterval, string headerName)
+        {
+            var originalDelay = TimeSpan.FromMilliseconds(500);
+            var shortDelay = originalDelay.Subtract(TimeSpan.FromMilliseconds(250));
+            var longDelay = originalDelay.Add(TimeSpan.FromMilliseconds(250));
+            var passedDelays = new List<TimeSpan>();
+
+            var shortDelayMockResponse = new MockResponse(200);
+            shortDelayMockResponse.AddHeader(new HttpHeader(headerName, shortDelay.Milliseconds.ToString()));
+
+            var longDelayMockResponse = new MockResponse(200);
+            longDelayMockResponse.AddHeader(new HttpHeader(headerName, longDelay.Milliseconds.ToString()));
+
+            var callsCount = 0;
+            var testOperation = new TestOperation()
+            {
+                OnUpdateState = _ => ++callsCount switch
+                {
+                    1 => OperationState<int>.Pending(shortDelayMockResponse),
+                    2 => OperationState<int>.Pending(longDelayMockResponse),
+                    _ => OperationState<int>.Success(new MockResponse(200), 50)
+                }
+            };
+            var operationInternal = testOperation.MockOperationInternal;
+
+            operationInternal.OnWait = delay => passedDelays.Add(delay);
+
+            if (useDefaultPollingInterval)
+            {
+                operationInternal.DefaultPollingInterval = originalDelay;
+                await operationInternal.WaitForCompletionAsync(CancellationToken.None);
+            }
+            else
+            {
+                await operationInternal.WaitForCompletionAsync(originalDelay, CancellationToken.None);
+            }
+
+            // Algorithm must choose the longest delay between the two.
```

Comment:
Perhaps I am missing it, but it doesn't look like we assert this anywhere.

Reply:
We define three delays: `originalDelay`, `shortDelay`, and `longDelay`.
https://github.com/Azure/azure-sdk-for-net/blob/c24a61f1bf175f9723bf5868fc8141c96ecb088e/sdk/core/Azure.Core/tests/OperationInternalTests.cs#L396-L398
And we know that shortDelay < originalDelay < longDelay holds.

We set `originalDelay` to be the default polling time, then pass two mock responses containing the retry-after header to the operation.
1. In the first response, the header uses the short delay, so the algorithm must choose `originalDelay`.
2. In the second response, the header uses the long delay, so the algorithm must choose `longDelay`.

Reply:
I would  have expected to see an asserts that looks something like:

`Assert that delayResponse was x`
`Assert that default polling delay was y`
`Assert that value passed to WaitAsync was Math.Max(x, y)`